### PR TITLE
Change taint window caption dynamically

### PR DESCRIPTION
**Pull request by duncanp-sonar**
Relates to #2059

### CODE REVIEW

File: src/IssueViz.Security/Taint/TaintList/TaintToolWindow.cs [hunk 1]
```diff
@@ -52,7 +52,36 @@ public TaintToolWindow(IServiceProvider serviceProvider)
                 componentModel.GetService<ITelemetryManager>()
             );
 
-            Content = new TaintIssuesControl(viewModel);
+            Initialize(viewModel);
+        }
+
+        internal /* for testing */ TaintToolWindow(ITaintIssuesControlViewModel viewModel) =>
```
Comment (duncanp-sonar) [inline src/IssueViz.Security/Taint/TaintList/TaintToolWindow.cs:58]: I added a second constructor to make it easier to add some unit tests for the tool window class.

File: src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs [hunk 2]
```diff
@@ -177,9 +195,29 @@ private void UpdateIssues()
             NotifyPropertyChanged(nameof(HasServerIssues));
         }
 
+        private void UpdateCaption()
+        {
+            // We'll show the default caption if:
+            // * there are no underlying issues, or
+            // * there is not ana active document.
```
Comment (duncanp-sonar) [inline src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs:202]: Typo

**Comment by duncanp-sonar** [comment]
There are a couple of scenarios in which the caption will not update correctly. However, both of these are due to bugs in the `IActiveDocumentTracker`. I raised #2079 last week (single-click navigation), and found another one today (#2091 - filter not updated when the last document is closed).

File: src/IssueViz.Security.UnitTests/Taint/TaintList/TaintIssuesControlViewModelTests.Caption.cs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+
+namespace SonarLint.VisualStudio.IssueVisualization.Security.UnitTests.Taint.TaintList
+{
+    [TestClass]
+    public class TaintIssuesControlViewModelTestsCaption
+    {
+        private readonly string DefaultCaption = Resources.TaintToolWindowCaption;
+
+        [TestMethod]
+        public void Ctor_NoUnderlyingIssues_DefaultCaptionSet()
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/Taint/TaintList/TaintIssuesControlViewModelTests.Caption.cs:48]: There are four different cases to test to check that the correct caption is set when the window is first opened.

File: src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs [hunk 4]
```diff
@@ -177,9 +195,29 @@ private void UpdateIssues()
...
+            WindowCaption = Resources.TaintToolWindowCaption + suffix;
+        }
+
+        private int GetFilteredIssuesCount() =>
+            IssuesView.OfType<object>().Count();
+
         private void Store_IssuesChanged(object sender, IssuesChangedEventArgs e)
         {
             UpdateIssues();
+            UpdateCaption();
```
Comment (rita-g-sonarsource) [inline src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs:220]: Minor: why not set the UpdateCaption inside UpdateIssues? This way we won't need to explicitly call it from the ctor as well.

File: src/IssueViz.Security.UnitTests/Taint/TaintList/TaintIssuesControlViewModelTests.Caption.cs [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+        {
+            var testCase = new TestCase()
```
Comment (rita-g-sonarsource) [inline src/IssueViz.Security.UnitTests/Taint/TaintList/TaintIssuesControlViewModelTests.Caption.cs:50]: Very neat syntax, I like it.

**Review by rita-g-sonarsource — APPROVED** [review]

File: src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs (hunk 4 above)
Comment (duncanp-sonar) [reply src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs:220]: That's what I did originally but it's not enough on its own - the caption needs to be set after the issues have been set *and* the filter has been initialised.
In the end it was simpler just to make explicit calls in a couple of places.

File: src/IssueViz.Security.UnitTests/Taint/TaintList/TaintIssuesControlViewModelTests.Caption.cs (hunk 5 above)
Comment (duncanp-sonar) [reply src/IssueViz.Security.UnitTests/Taint/TaintList/TaintIssuesControlViewModelTests.Caption.cs:50]: Yep, it is similar; the main difference apart from readability is test isolation.

**Pull request merged**